internal/cmd/temporal-features-test-proxy/main.go

```diff
@@ -0,0 +1,752 @@
+		return fmt.Errorf("failed to listen on %q: %w", s.listen, err)
+	}
+
+	needListenerClose := true
+	defer func() {
+		if needListenerClose {
+			err := l.Close()
+			if IsErrClosed(err) {
+				err = nil
+			}
+			if err != nil {
+				Warn("failed to close listener for gRPC proxy server: %v", err)
+			}
+		}
+	}()
+
+	gc, err := grpc.DialContext(
+		ctx,
+		s.dial,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithBlock(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to dial %q: %w", s.dial, err)
+	}
+
+	needClientClose := true
+	defer func() {
+		if needClientClose {
+			err := gc.Close()
+			if IsErrClosed(err) {
+				err = nil
+			}
+			if err != nil {
+				Warn("failed to close gRPC client connection: %v", err)
+			}
+		}
+	}()
+
+	wc := workflowservice.NewWorkflowServiceClient(gc)
+	ws, err := client.NewWorkflowServiceProxyServer(client.WorkflowServiceProxyOptions{Client: wc})
+	if err != nil {
+		return fmt.Errorf("failed to create WorkflowService proxy server: %w", err)
+	}
+
+	gs := grpc.NewServer(
+		grpc.UnaryInterceptor(ProxyUnaryInterceptor),
+		grpc.StreamInterceptor(ProxyStreamInterceptor),
+	)
+	grpc_health_v1.RegisterHealthServer(gs, &TrivialHealthServer{})
+	workflowservice.RegisterWorkflowServiceServer(gs, ws)
+
+	needClientClose = false
+	needListenerClose = false
+	s.gc = gc
+	s.gs = gs
+	s.l = l
+	s.wc = wc
+	s.ws = ws
+	s.quitCh = make(chan struct{})
+
+	go s.serveThread()
+	go s.closeThread(ctx, s.quitCh, s.gs)
+	return nil
+}
+
+func (s *ProxyServer) Shutdown(ctx context.Context) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.quitCh == nil {
+		return nil
+	}
+
+	close(s.quitCh)
+	s.gs.GracefulStop()
+	for s.quitCh != nil {
+		s.cv.Wait()
+	}
+	return nil
+}
+
+func (s *ProxyServer) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.quitCh == nil {
+		return nil
+	}
+
+	close(s.quitCh)
+	s.gs.Stop()
+	for s.quitCh != nil {
+		s.cv.Wait()
+	}
+	return nil
+}
+
+func (s *ProxyServer) ForceClose() {
+	err := s.Close()
+	if IsErrClosed(err) {
+		err = nil
+	}
+	if err != nil {
+		Warn("failed to stop gRPC proxy server: %v", err)
+	}
+}
+
+type Stopper interface {
+	Stop()
+}
+
+func (s *ProxyServer) closeThread(ctx context.Context, quitCh <-chan struct{}, stopper Stopper) {
+	select {
+	case <-ctx.Done():
+		stopper.Stop()
+	case <-quitCh:
+	}
+}
+
+func (s *ProxyServer) serveThread() {
+	defer s.finish()
+
+	err := s.gs.Serve(s.l)
+	if IsErrClosed(err) {
+		err = nil
+	}
+	if err != nil {
+		Error("failed to serve gRPC proxy server: %v", err)
+	}
+
+	err = s.l.Close()
+	if IsErrClosed(err) {
+		err = nil
+	}
+	if err != nil {
+		Warn("failed to close listener for gRPC proxy server: %v", err)
+	}
+
+	err = s.gc.Close()
+	if IsErrClosed(err) {
+		err = nil
+	}
+	if err != nil {
+		Warn("failed to close gRPC client connection: %v", err)
+	}
+}
+
+func (s *ProxyServer) finish() {
+	s.mu.Lock()
+	s.gc = nil
+	s.gs = nil
+	s.l = nil
+	s.wc = nil
+	s.ws = nil
+	s.quitCh = nil
```

Comment:
Seems like it might be easy to miss any future fields here. I'm wondering if just having three top-level fields: The mutex, condvar, and an "everything else" container might be nice.

It could also potentially let you have some generic code for this pattern that the servers can share.